### ISSUE TITLE
External secrets store

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#17257]
## Overview

Materialize today has built-in support for secrets using [`CREATE SECRET`](https://materialize.com/docs/sql/create-secret/). These secrets are stored securely in Materialize's cloud environment.

We've had at least one request to integrate with external secret store ([internal Slack](https://materializeinc.slack.com/archives/CU7ELJ6E9/p1672656260325249)). 

The two most common secret stores that I know of are:

  * [HashiCorp Vault](https://developer.hashicorp.com/vault/docs/what-is-vault)
  * [AWS Secrets Manager](https://docs.aws.amazon.com/secretsmanager/index.html)

## Motivation

Integrating with an external secret store provides only a marginal security benefit. You still have to trust Materialize to handle your secret credentials. 

I dp see two small benefits to the integration:

  1. Organizations that already use an external store to manage their secrets don't need to manually copy secrets into Materialize and/or write a custom sync between their external secrets store and Materialize. 
  2. In the case of a breach, secrets could be revoked from Materialize centrally in the external secrets store, instead of requiring someone to log into Materialize to manually run `DROP SECRET`.

## Design questions

* How do we represent a connection to the secrets store in the catalog? `CREATE CONNECTION`?
* How do you refer to a secret in an external secrets store? `SECRET connection.name_of_secret_in_external_store`?
* Once we have RBAC, does RBAC apply to the connection as a whole or to individual secrets in the external store?

## Timeline

We haven't seen much user interest in external secret stores yet. I don't imagine we'd build this for several years.

Please leave a thumbs up if you're interested in this feature. Better yet, leave a comment explaining your use case.

**Comment by morsapaes** [comment]
When the time comes (if it ever does), could we extend the `CREATE SECRET` syntax to also allow a `FROM secret_store_type CONNECTION ( options )` clause? This would be more consistent with existing syntax for other objects that use connections to external services. I imagine we'll have a more ergonomic interface than raw DDL for connection and secret management by then, in any case.

**Comment by benesch** [comment]
> When the time comes (if it ever does), could we extend the `CREATE SECRET` syntax to also allow a `FROM secret_store_type CONNECTION ( options )` clause? This would be more consistent with existing syntax for other objects that use connections to external services. I imagine we'll have a more ergonomic interface than raw DDL for connection and secret management by then, in any case.

Yes, in theory, but I feel like the big benefit is that you don't have to run `CREATE SECRET` individually for each secret in the external store? Like, you'd just connect to your HashiCorp vault once and then you'd have access to all the secrets as you liked:

```
CREATE CONNECTION hashicorp_vault TO HASHICORP VAULT (
    HOST '...',
);

-- No CREATE SECRET required!

CREATE CONNECTION kafka TO KAFKA (
    -- ...
    SASL USER hashicorp_vault.my_kafka_sasl_user
    SASL PASSWORD hashicorp_vault.my_kafka_sasl_password
);
```

Am I misunderstanding what you meant?

**Comment by morsapaes** [comment]
You're not! I just defaulted to thinking that you'd still want to explicitly create and control access to individual secrets from the external store as Materialize objects (and mainly avoid handling raw credentials), but it does defeat the whole purpose. 🫠

**Comment by benesch** [comment]
Well, it defeats one plausible purpose! But we're still speculating as to why users might want to use an external secrets store with Materialize. Could well be that there are use cases for having it work as you describe.

**Comment by morsapaes** [comment]
In the meantime, one thing we discussed in the DevEx team was integrating with external secret stores via Terraform. @bobbyiliev put together an example using HashiCorp Vault [here](https://github.com/MaterializeInc/demos/pull/76).
